functions/src/notifications.ts

```diff
@@ -92,6 +92,75 @@ export function createActivityNotification(token: string, activity_id: string):
     }
 }
 
+// Sends a message to a chat if a user leaves or joins it:
+export async function onActivityUsersChanged(change: Change<FirebaseFirestore.DocumentSnapshot>, context: EventContext) {
+    const activityId: string = change.after.id
+    const usersBefore = getUsersFromChange(change.before)
+    const usersAfter = getUsersFromChange(change.after)
+
+    if (usersBefore.length !== usersAfter.length) {
+        // Gets which user joined or left the activity. changedUser will be that person's UID.
+        const changedUser = getDifferentString(usersBefore, usersAfter)
+        if (! changedUser) return -1
+
+        // @ts-ignore because it doesn't like the mock
+        const userInfo = await Refs(database).user(changedUser).get()
+        if (userInfo.exists) {
+            const changedUserInfo = userInfo.data()
+            // Setup message based on whether they left or joined:
+            const msgBody = `${changedUserInfo.name} has ${usersBefore > usersAfter ? 'left' : 'joined'} your activity.`
+            
+            return exports.sendChatMessage(activityId, msgBody, changedUser, new Date())
+        }
+    }
+    return 0
+}
+
+// Sends a user left/joined message to the specified activity
+export function sendChatMessage(activityId: string, message: string, sender: string, date: Date) {
```

Comment:
Is this function supposed to be generic or _only_ for join/leave?  I'm assuming the `type` won't be hard-coded eventually?

Reply:
That's true, we wouldn't have to encode the type in the function, though at this point we don't have plans for the function to do anything else :) I think this case (since it's so simple) would be fine to leave until we do need to repeat ourselves